interp/interp.go

```diff
@@ -0,0 +1,158 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package interp
+
+import (
+	"sort"
+)
+
+// Predictor predicts the value of a function. It handles both
+// interpolation and extrapolation.
+type Predictor interface {
```

Comment:
I'm not convinced by the name because to me "predict" has a temporal meaning but I don't have a better suggestion other than the obvious Interpolator. On the other hand, the interface doesn't transmit anything related to interpolation apart from the name and being in the `interp` package. After the interpolator is created it behaves just as a regular function with the restriction that values outside of the interpolation region are undefined or defined arbitrarily. Perhaps this could be reflected in the interface by adding a function that provides the region of validity?

Reply:
See here: https://stats.stackexchange.com/questions/205352/what-is-the-exact-difference-between-prediction-and-extrapolation. I think `Predict` is a good name for a function which does both interpolation and extrapolation.

There was an `Interval` function in the previous version of the PR, but it was removed following previous reviews.